metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,423 @@
+          indentation: String,
+          path,
+          termIf.pos.toLSP.getStart,
+          termIf.pos.toLSP.getEnd
+        )
+      )
+    else None
+  }
+
+  private def buildCodeActionFromForYieldParts(
+      nameQualsList: List[ForYieldEnumeration],
+      forYieldConditionsList: List[ForYieldCondition],
+      yieldExpression: YieldExpression,
+      indentation: String,
+      path: AbsolutePath,
+      startPos: l.Position,
+      endPos: l.Position
+  ): l.CodeAction = {
+    val forYieldString =
+      s"""|for {
+          |${nameQualsList.map(nameQual => s"${indentation}  ${nameQual.variableName} <- ${nameQual.qual}").mkString("\n")}
+          |${forYieldConditionsList.map(forYieldCondition => s"${indentation}  if ${forYieldCondition.condition}").mkString("\n")}
+          |${indentation}} yield {
+          |${indentation}  ${yieldExpression.expression}
+          |${indentation}}""".stripMargin
+
+    val codeAction = new l.CodeAction()
+    val range =
+      new l.Range(startPos, endPos)
+    codeAction.setTitle(
+      FlatMapToForComprehensionCodeAction.flatMapToForComprehension
+    )
+    codeAction.setKind(this.kind)
+    val forComprehensionTextEdit = new l.TextEdit(range, forYieldString)
+    codeAction.setEdit(
+      new l.WorkspaceEdit(
+        Map(
+          path.toURI.toString -> List(forComprehensionTextEdit).asJava
+        ).asJava
+      )
+    )
+    codeAction
+  }
+
+  private def maybeBuildToForYieldCodeActionWithApply(
+      path: AbsolutePath,
+      termApply: Term.Apply,
+      document: String,
+      indentation: String
+  ): Option[l.CodeAction] = {
+    val (nameQualsList, conditions, yieldExpression) = {
+      extractForYield(List.empty, List.empty, termApply, document)
+    }
+
+    if (nameQualsList.nonEmpty) {
+      Some(
+        buildCodeActionFromForYieldParts(
+          nameQualsList,
+          conditions,
+          yieldExpression,
+          indentation,
+          path,
+          termApply.pos.toLSP.getStart,
+          termApply.pos.toLSP.getEnd
+        )
+      )
+    } else None
+
+  }
+
+  private def maybeBuildToForYieldCodeActionWithSelect(
+      path: AbsolutePath,
+      termSelect: Term.Select,
+      document: String,
+      indentation: String
+  ): Option[l.CodeAction] = {
+    termSelect.parent.collect { case termApply: Term.Apply =>
+      maybeBuildToForYieldCodeActionWithApply(
+        path,
+        termApply,
+        document,
+        indentation
+      )
+    }.flatten
+  }
+
+  private def maybeBuildToForYieldCodeActionWithName(
+      path: AbsolutePath,
+      termName: Term.Name,
+      document: String,
+      indentation: String
+  ): Option[l.CodeAction] = {
+    termName.parent.collect { case termSelect: Term.Select =>
+      maybeBuildToForYieldCodeActionWithSelect(
+        path,
+        termSelect,
+        document,
+        indentation
+      )
+    }.flatten
+
+  }
+
+  case class ForYieldEnumeration(variableName: String, qual: String)
```

Comment:
There is already Enumeration Tree type that can be used for this purpose and it already has all the correct sub typing.

Reply:
Interesting! I did not pay attention to that!

Reply:
But `Enumeration`tree type is much more heftier than just a variable name and qual string I think. Since I am not constructing the tree. I just write the strings of those enumeration in the deducted TextEdit.

Reply:
The main advantage is that you will be able to print the Tree afterwards and be sure that it's syntactically correct.

Reply:
Ok, we have `Enumerator.Generator` and `Enumerator.Val` for this. So the main fields which are to be populated are `pat` for the variable name and `rhs` for the right hand side of the arrow. 
Now the question is this:
how to generate the `pos` field of the trees which are going to get assigned  to the `pat` and `rhs` fields?
and more specifically, what numbers to assign to the start and end values.
Also, these are the methods that should get implemented when manually creating an instance of `Enumerator.Generaor`
```
  override def copy(pat: Pat, rhs: Term): Enumerator.Generator = ???

      override def children: List[Tree] = ???

      override private[meta] def privatePrototype = ???

      override private[meta] def privateParent = ???

      override private[meta] def privateOrigin = ???

      override private[meta] def privateCopy(prototype: Tree, parent: Tree, destination: String, origin: Origin) = ???

      override def productFields: List[String] = ???

      override def productArity: Int = ???

      override def productElement(n: Int): Any = ???
```

The other question is how to implement them?

Is there a more straightforward way for instantiating `Enumerator.Generator` and `Enumerator.Val` which does not require that much work?